### PR TITLE
Use `float` instead of `Date` for protocol types for higher precision.

### LINKED ISSUE #1689: Use `float` instead of `Date` for protocol types for higher precision
* If possible add an overload to with `float` but keep the `Date` one
* add backward compatibility when deserializing payloads

The goal is to have a higher precision in the performance product in particular. So types such as Span and Transaction should have `float` but it's less of a concern to something like Breadcrumb

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :scroll: Description

Use `float` instead of `Date` for protocol types for higher precision.

To retrieve more precise timestamps, we are using `System#nanoTime` as it's the only method available both on Android and JVM. 

When Span is created with given start timestamp (for example with Android app start time), we are not able to measure high precision timestamp, we fall back then to regular timestamp with milliseconds precision.

How it looks in UI:
image

## :bulb: Motivation and Context

Fixes #1689

## :green_heart: How did you test it?

Unit & integration tests.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

File: sentry/src/main/java/io/sentry/Span.java
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/Span.java:105]: Is there a way to make this change without a breaking change?
Like adding another `timestampHigherPrecision()` and leave this one as `Date timestamp()` and have it return a `Date` from the float for example?

We'd need to bump a major to release this change

**Review by bruno-garcia — COMMENTED** [review]
Looks so much better in the UI 💯

File: sentry/src/main/java/io/sentry/Span.java
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/Span.java:105]: Definitely there is a way, the question is if it's worth it considering this class is marked as internal. `getTimestamp` is never used by the user code as it's not exposed with `ISpan`.

**Comment by marandaneto** [comment]
btw: `System#nanoTime` 

>  This clock is guaranteed
>  *     to be monotonic, and is suitable for interval timing when the
>  *     interval does not span device sleep

that means when the device is in sleep mode, the system does not increase the timer, is it possible that it'd cause off-by-1 sort of problems?

File: sentry/src/main/java/io/sentry/adapters/DoubleDeserializerAdapter.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/adapters/DoubleDeserializerAdapter.java:17]: does that mean that this class and `DoubleSerializerAdapter` will be triggered for every Double in the protocol classes and if its not related to the Date/Double changes?

File: sentry/src/main/java/io/sentry/adapters/DoubleSerializerAdapter.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/adapters/DoubleSerializerAdapter.java:33]: if we try to serialize a Double that is added by setExtra or setContexts (not related to the Date/Double changes), is it gonna try to scale down?

File: sentry/src/main/java/io/sentry/adapters/DoubleDeserializerAdapter.java
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/adapters/DoubleDeserializerAdapter.java:17]: Yes, and I think there's no easy way to set up an adapter only for fields related to one type - we should I think have a separate serializer for `SentryTransaction` and `SentrySpan`. Perhaps its better to wait for the non-gson serializer to be merged till we merge this one?

File: sentry/src/main/java/io/sentry/adapters/DoubleSerializerAdapter.java
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/adapters/DoubleSerializerAdapter.java:33]: Good spot. Look comment [above](https://github.com/getsentry/sentry-java/pull/1737#discussion_r716774333)

**Comment by maciejwalkowiak** [comment]
> that means when the device is in sleep mode, the system does not increase the timer, is it possible that it'd cause off-by-1 sort of problems?

Are our transactions meant to span over the system sleep? If so, we should at least check if the `endTime` - `startTime` isn't a negative number.

File: sentry/src/main/java/io/sentry/adapters/DoubleDeserializerAdapter.java
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/adapters/DoubleDeserializerAdapter.java:17]: yeah I'd say so, so we can target the ser/deser correctly

**Comment by marandaneto** [comment]
> > that means when the device is in sleep mode, the system does not increase the timer, is it possible that it'd cause off-by-1 sort of problems?
> 
> Are our transactions meant to span over the system sleep? If so, we should at least check if the `endTime` - `startTime` isn't a negative number.

good question and I don't know the answer right now, have to check if it makes sense or not

**Comment by marandaneto** [comment]
should we target the v6 branch then?

**Comment by maciejwalkowiak** [comment]
Yes but this PR is anyway on hold until we have JSON serialization merged.

**Comment by bruno-garcia** [comment]
Gson PR merged into branch `6.x.x`. Lets get this one merged in there too? We'll release an alpha soon.

**Comment by maciejwalkowiak** [comment]
6.x.x is now up to date with main, this PR is now pointing to 6.x.x. Ready for the final review.

**Comment by marandaneto** [comment]
@romtsn would you like to have a look?

**Comment by romtsn** [comment]
@marandaneto checking 👍

**Review by romtsn — APPROVED** [review]
LGTM

**Review by marandaneto — APPROVED** [review]

**Pull request merged**
